src/block_storage/volumes.rs

```diff
@@ -0,0 +1,453 @@
+use super::{api, protocol};
+
+/// A query to volume list.
+#[derive(Clone, Debug)]
+pub struct VolumeQuery {
+    session: Session,
+    query: Query,
+    can_paginate: bool,
+    sort: Vec<String>,
+}
+
+/// Structure representing a summary of a single volume.
+#[derive(Clone, Debug)]
+pub struct Volume {
+    session: Session,
+    inner: protocol::Volume,
+}
+
+/// A request to create a volume.
+#[derive(Clone, Debug)]
+pub struct NewVolume {
+    session: Session,
+    inner: protocol::VolumeCreate,
+}
+
+impl Display for Volume {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{:#?}", self.inner)
+    }
+}
+
+impl Volume {
+    /// Create an Volume object.
+    pub(crate) async fn new<Id: AsRef<str>>(session: Session, id: Id) -> Result<Volume> {
+        let inner = api::get_volume(&session, id).await?;
+        Ok(Volume { session, inner })
+    }
+
+    transparent_property! {
+        #[doc = "Migration status."]
+        migration_status: ref Option<String>
+    }
+
+    transparent_property! {
+        #[doc = "Volume attachments."]
+        attachments: ref Vec<protocol::VolumeAttachment>
+    }
+
+    transparent_property! {
+        #[doc = "Volume links."]
+        links: ref Vec<protocol::Link>
+    }
+
+    transparent_property! {
+        #[doc = "Name of the availability zone."]
+        availability_zone: ref Option<String>
+    }
+
+    transparent_property! {
+        #[doc = "Current backend of the volume."]
+        host: ref Option<String>
+    }
+
+    transparent_property! {
+        #[doc = "Whether the volume is encrypted."]
+        encrypted: ref bool
+    }
+
+    transparent_property! {
+        #[doc = "UUID of the encryption key."]
+        encryption_key_id: ref Option<String>
+    }
+
+    transparent_property! {
+        #[doc = "When the volume was last updated."]
+        updated_at: ref Option<String>
+    }
+
+    transparent_property! {
+        #[doc = "Volume replication status."]
+        replication_status: ref Option<String>
+    }
+
+    transparent_property! {
+        #[doc = "UUID of the snapshot the volume originated from."]
+        snapshot_id: ref Option<String>
+    }
+
+    transparent_property! {
+        #[doc = "UUID of the volume."]
+        id: ref String
+    }
+
+    transparent_property! {
+        #[doc = "Size of the volume in GiB."]
+        size: ref u64
+    }
+
+    transparent_property! {
+        #[doc = "UUID of the user."]
+        user_id: ref String
+    }
+
+    transparent_property! {
+        #[doc = "UUID of the project."]
+        tenant_id: ref Option<String>
+    }
+
+    transparent_property! {
+        #[doc = "Migration status."]
+        migstat: ref Option<String>
+    }
+
+    transparent_property! {
+        #[doc = "Metadata of the volume."]
+        metadata: ref HashMap<String, String>
+    }
+
+    transparent_property! {
+        #[doc = "Status of the volume."]
+        status: ref protocol::VolumeStatus
```

Comment:
Protocol enums are also Copy, you can omit ref here too.

Reply:
done